File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
      l, r = head, head
      while r and r.next:
        l = l.next
        r = r.next.next
      
      curr, prev = l, None
      while curr:
        next = curr.next
        curr.next = prev
        prev = curr
        curr = next
      
      while prev:
        if head.val == prev.val:
          head = head.next
          prev = prev.next
        else:
          return False
      
      return True
        

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    let l = head, r = head;
    while (r && r.next) {
      l = l.next
      r = r.next.next
    }

    let curr = l, prev = null
    while (curr) {
      let next = curr.next
      curr.next = prev
      prev = curr
      curr = next
    }
    
    while (prev) {
      if (prev.val === head.val) {
        prev = prev.next
        head = head.next
      } else {
        return false
      }
    }
    return true
};